### PR TITLE
Improve __repr__ for various objects

### PR DESCRIPTION
**Pull request by sarlinpe**
Breaking changes:
- `Sim3d.from_matrix` becomes a constructor `Sim3d(matrix)`
- `Camera.{model_id,model_name}` are merged into `Camera.model` (via enum)
- `Camera.params_info` and `Image.{num_points2D,num_points3D}` become read-only properties (previously: functions)

Samples:
```
>> camera
Camera(camera_id=1, model=SIMPLE_RADIAL, width=780, height=1063, params=[1245.950053, 390.000000, 531.500000, 0.020831] (f, cx, cy, k))

>> print(rec.cameras[1].summary())
Camera:
    camera_id = 1
    model = CameraModelId.SIMPLE_RADIAL
    width = 780
    height = 1063
    focal_length = 1245.950052502772
    focal_length_x = 1245.950052502772
    focal_length_y = 1245.950052502772
    has_prior_focal_length = False
    principal_point_x = 390.0
    principal_point_y = 531.5
    params_info = f, cx, cy, k
    params = [1.24595005e+03 3.90000000e+02 5.31500000e+02 2.08308406e-02]

>> image
Image(image_id=1, camera_id=1, name="mapping/02928139_3448003521.jpg", triangulated=760/2313)

>> print(image.summary())
Image:
    image_id = 1
    camera_id = 1
    name = mapping/02928139_3448003521.jpg
    cam_from_world = Rigid3d(quat_xyzw=[0.0500042, 0.243753, -0.155877, 0.955922], t=[-1.37836, 0.000488448, -1.59398])
    cam_from_world_prior = Rigid3d(quat_xyzw=[nan, nan, nan, nan], t=[nan, nan, nan])
    points2D = [ ... 2313 elements ... ]
    registered = True
    num_points2D = 2313
    num_points3D = 760
    num_observations = 0
    num_correspondences = 0

>> sim3d
Sim3d(scale=1, quat_xyzw=[0, 0, 0, 1], t=[0, 0, 0])
```

### CODE REVIEW

File: pycolmap/scene/image.h [hunk 1]
```diff
@@ -286,19 +287,5 @@ void BindImage(py::module& m) {
...
-                   : "Invalid")
-           << "\n\tcamera_id = "
-           << (self.HasCamera() ? std::to_string(self.CameraId()) : "Invalid")
-           << "\n\tname = " << self.Name()
-           << "\n\ttriangulated = " << self.NumPoints3D() << "/"
-           << self.NumPoints2D() << "\n\tcam_from_world = ["
-           << self.CamFromWorld().ToMatrix() << "]";
-        return ss.str();
-      });
+      .def("__repr__", &PrintImage);
```
Comment (Phil26AT) [inline pycolmap/scene/image.h:290]: I think there should be a way to print all the parameters of an image, including the pose. I suggest to either bring the summary back (my preferred option), or add the pose to `__repr__`.

File: pycolmap/sfm/incremental_triangulator.h [hunk 2]
```diff
@@ -98,7 +98,7 @@ void BindIncrementalTriangulator(py::module& m) {
            })
       .def("__repr__", [](const IncrementalTriangulator& self) {
         std::stringstream ss;
-        ss << "<IncrementalTriangulator>";
+        ss << "IncrementalTriangulator()";
```
Comment (Phil26AT) [inline pycolmap/sfm/incremental_triangulator.h:101]: This is quite obscure and useless. Maybe print the underlying reconstruction and correspondence graph instead?

File: pycolmap/geometry/bindings.h [hunk 3]
```diff
@@ -30,9 +31,10 @@ void BindGeometry(py::module& m) {
       .def("inverse", &Eigen::Quaterniond::inverse)
       .def("__repr__", [](const Eigen::Quaterniond& self) {
         std::stringstream ss;
-        ss << "Rotation3d: " << self.coeffs();
+        ss << "Rotation3d(quat_xyzw=[" << self.coeffs().transpose() << "])";
         return ss.str();
       });
+  py::implicitly_convertible<py::array, Eigen::Quaterniond>();
```
Comment (Phil26AT) [inline pycolmap/geometry/bindings.h:37]: That's a nice idea, maybe also add a conversion from a 3x4 matrix directly to Sim3d / Rigid3d?

File: pycolmap/scene/camera.h [hunk 4]
```diff
@@ -184,24 +200,11 @@ void BindCamera(py::module& m) {
       .def("__copy__", [](const Camera& self) { return Camera(self); })
       .def("__deepcopy__",
            [](const Camera& self, py::dict) { return Camera(self); })
-      .def("__repr__", [](const Camera& self) { return PrintCamera(self); })
-      .def("summary", [](const Camera& self) {
```
Comment (Phil26AT) [inline pycolmap/scene/camera.h:188]: It is a bit annoying to have to look up which classes have `.summary`. Maybe add it to all classes, and by default just redirect to `__repr__`? `__repr__` for a very short string identifier, `.summary` for a longer, more detailed description.

File: pycolmap/scene/camera.h [hunk 5]
```diff
@@ -44,7 +46,21 @@ void BindCamera(py::module& m) {
   AddStringToEnumConstructor(PyCameraModelId);
   py::implicitly_convertible<int, CameraModelId>();
 
-  py::bind_map<CameraMap>(m, "MapCameraIdCamera");
+  py::bind_map<CameraMap>(m, "MapCamerasById")
```
[the remark this answers is not in the document]
Comment (sarlinpe) [reply pycolmap/scene/camera.h]: Agree, changed again.

File: pycolmap/scene/camera.h [hunk 6]
```diff
@@ -61,8 +77,8 @@ void BindCamera(py::module& m) {
       .def_property(
           "model_name",
           &Camera::ModelName,
-          [](Camera& self, std::string model_name) {
-            self.model_id = CameraModelNameToId(model_name);
+          [](Camera& self, CameraModelId model_id) {  // implicit conversion
```
[the remark this answers is not in the document]
Comment (sarlinpe) [reply pycolmap/scene/camera.h]: Good point, now the interface is much simpler. The string name can be recovered with `camera.model.name`.

File: pycolmap/scene/image.h (hunk 1 above)
Comment (sarlinpe) [reply pycolmap/scene/image.h:294]: Made Image a dataclass to automatically generate the summary function.

File: pycolmap/scene/camera.h (hunk 4 above)
Comment (sarlinpe) [reply pycolmap/scene/camera.h:188]: Now we generate the summary automatically.

File: pycolmap/scene/camera.h [hunk 7]
```diff
@@ -19,15 +19,17 @@ namespace py = pybind11;
...
   std::stringstream ss;
-  ss << "<Camera 'camera_id="
+  ss << "Camera(camera_id="
      << (camera.camera_id != kInvalidCameraId ? std::to_string(camera.camera_id)
                                               : "Invalid")
-     << ", model=" << camera.ModelName() << ", width=" << camera.width
-     << ", height=" << camera.height << ", num_params=" << camera.params.size()
-     << "'>";
+     << ", model=" << model_name << ", width=" << camera.width
+     << ", height=" << camera.height << ", params=[" << camera.ParamsToString()
```
[the remark this answers is not in the document]
Comment (Phil26AT) [reply pycolmap/scene/camera.h:34]: Makes sense

File: pycolmap/scene/image.h (hunk 1 above)
Comment (Phil26AT) [reply pycolmap/scene/image.h:294]: Great!

**Pull request merged**